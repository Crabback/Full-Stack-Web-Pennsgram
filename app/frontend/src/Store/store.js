/* istanbul ignore file */

import { configureStore } from '@reduxjs/toolkit'
import currentUserReducer from '../pages/UserPage/currentUserSlice'

export default configureStore({
  reducer: {
    currentUser: currentUserReducer
  }
})