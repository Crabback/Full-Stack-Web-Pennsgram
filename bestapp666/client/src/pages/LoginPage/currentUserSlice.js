import { createSlice } from '@reduxjs/toolkit'

export const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState: {
    id: -1,
    username: '',
    password: '',
    followings: [],
    posts: []
  },
  reducers: {
    addLoginUser: (state,action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state =  action.user
    },
    logoutAction: state => {
      state = {
        id: -1,
        username: '',
        password: '',
        followings: [],
        posts: []
      }
    },
  }
})

export const { addLoginUser, logoutAction } = currentUserSlice.actions

export default currentUserSlice.reducer