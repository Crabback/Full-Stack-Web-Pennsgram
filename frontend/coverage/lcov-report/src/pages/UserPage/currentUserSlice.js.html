
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/pages/UserPage/currentUserSlice.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> / <a href="index.html">src/pages/UserPage</a> currentUserSlice.js</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>0/6</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/0</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/3</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>0/5</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { createSlice } from '@reduxjs/toolkit'
&nbsp;
export const currentUserSlice = <span class="cstat-no" title="statement not covered" >createSlice({</span>
  name: 'currentUser',
  initialState: {
    id: -1,
    username: 'NOT_A_USER',
    password: '',
    followings: [],
    followers: [],
    avatar:"https://img.wattpad.com/8f19b412f2223afe4288ed0904120a48b7a38ce1/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f776174747061642d6d656469612d736572766963652f53746f7279496d6167652f5650722d38464e2d744a515349673d3d2d3234323931353831302e313434336539633161633764383437652e6a7067",
    posts: []
  },
  reducers: {
    updateCurrentUser: <span class="fstat-no" title="function not covered" >(s</span>tate, action) =&gt; {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
<span class="cstat-no" title="statement not covered" >      return action.payload </span>//you must return payload to receive the 
    },
    logoutAction: <span class="fstat-no" title="function not covered" >st</span>ate =&gt; {
<span class="cstat-no" title="statement not covered" >      return state = {</span>
        id: -1,
        username: 'NOT_A_USER',
        password: '',
        followings: [],
        followers: [],
        avatar:"https://img.wattpad.com/8f19b412f2223afe4288ed0904120a48b7a38ce1/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f776174747061642d6d656469612d736572766963652f53746f7279496d6167652f5650722d38464e2d744a515349673d3d2d3234323931353831302e313434336539633161633764383437652e6a7067",
        posts: []
      }
    },
  }
})
&nbsp;
export const { updateCurrentUser, logoutAction } = <span class="cstat-no" title="statement not covered" >currentUserSlice.actions</span>
&nbsp;
// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) =&gt; state.counter.value)`
export const selectCurrentUser = <span class="cstat-no" title="statement not covered" ><span class="fstat-no" title="function not covered" >(s</span>tate) =&gt; <span class="cstat-no" title="statement not covered" >state.currentUser</span></span>
&nbsp;
export default currentUserSlice.reducer</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2022-10-21T19:46:08.902Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    