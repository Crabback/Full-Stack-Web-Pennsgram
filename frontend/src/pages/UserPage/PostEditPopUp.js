import React from "react";
import popupStyles from "./post-popup.module.css";
import { useState, useEffect } from "react";
import PropTypes from "prop-types";
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import Toast from 'react-bootstrap/Toast';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import {deletePost, getComments, deleteComment, updatePost} from "../../fetcher";
import { NavLink } from "react-router-dom";

const PostEditPopUp = (props) => {

    const [show, setShow] = useState(false);
    const [media, setMedia] = useState("");
    const [descInput, setDescInput] = useState("");

    const doneHandler = (e) => {
        // update the database and dispatch currentUser
        //remember set descInput
        setShow(false);
        props.onClose(false);
    };

    const closeHandler = (e) => {
        setShow(false);
        setDescInput("");
        props.onClose(false);
    };

    useEffect(() => {
        setShow(props.show);
        setMedia(props.post.image);
        setDescInput(props.post.description);
        console.log("current post: ", props.post);
        }, [props.show]);
        
    const handleDescInput = (e) =>{
        setDescInput( e.target.value);
    };

    const handleImageURLInput = (e) => {
        setMedia(e.target.value);
      }
    
    // not finished
    async function handleDeleteComment(postId, author, content){
        try{
            await deleteComment(postId, author, content);
        }catch(err){
            console.log("Delete comment failed.");
            alert("Delete comment failed.");
        }
        //setEditedAndRefreshCards to trigger rerender and data fetching
        props.setHearFromDeleteComment(!props.oldHearFromDeleteComment); //pull the trigger 
    };

    async function handleDeletePost(postId){
        try{
            await deletePost(postId);
            // empty the parent's postBeingEdited in the edit window or will be error when delete;
            props.setPostBeingEdited({});
        }catch(err){
            console.log("delete post failed, id:", postId);
            alert("Delete post failed.");
        }
        //setEditedAndRefreshCards to trigger rerender and data fetching
        props.setEditedAndRefreshCards(!props.oldEditedAndRefreshCards);
        setShow(false);
        props.onClose(false);
    };

    const handleUpdate = async (e) => {
        e.preventDefault();
        try{
            await updatePost(props.post.id, media, descInput);
        }catch(err){
            console.log("update post image/caption failed");
            alert("update post image/caption failed");
        }
        //setEditedAndRefreshCards to trigger rerender and data fetching
        props.setEditedAndRefreshCards(!props.oldEditedAndRefreshCards);
        setShow(false);
        props.onClose(false);
    };

    function addDefaultImgSrc(ev){
        ev.target.src = "https://st4.depositphotos.com/14953852/24787/v/600/depositphotos_247872612-stock-illustration-no-image-available-icon-vector.jpg";
        setMedia(props.post.image);
      }

    function MediaPreview(props){
        if (props.mediaLink){
            if (props.mediaLink.split(".").slice(-1) == 'mp4'){
                return (
                <Form.Group className="mb-3">
                <video width="500" controls >
                    <source src={props.mediaLink} type="video/mp4"/>
                </video>
                </Form.Group>
                ) 
            }else{
                return (
                <Form.Group className="mb-3">
                    <img onError={addDefaultImgSrc} src={props.mediaLink} alt="user pic 1" width="400" height="400"></img>
                </Form.Group>
                )
            }
        }else{
            return (
                <Form.Group className="mb-3">
                    <img onError={addDefaultImgSrc} src={media} alt="user pic 1" width="400" height="400"></img>
                </Form.Group>
                )
        }
    }

    const commentsToast = (props.post.comments?  props.post.comments : [{author: "Void",
                                                    comment: "No comments",
                                                    mention: "@[NOT_A_USER](NOT_A_USER)"}]
    ).map((p, idx) => {
        return (
        <Toast show={true} onClose={(e) => handleDeleteComment(props.post.id, p.author, p.comment)} delay={1000}  className="d-inline-block m-1" bg={"light"} key={idx}>
            <Toast.Header>
                <img src="holder.js/20x20?text=%20" className="rounded me-2" alt=""/>
                <strong className="me-auto">{p.author}</strong>
                <small>11 mins ago</small>
            </Toast.Header>
            <Toast.Body >
                {p.comment + " "} 
                <NavLink to={"/user/"+p.mention.replace(')', '').split("(")[1]} className="button_text">
                    {p.mention.split("(")[0]}
                </NavLink>
            </Toast.Body>
        </Toast>
    )});

    return (
        <div style={{visibility: show ? "visible" : "hidden", opacity: show ? "1" : "0"}} className={popupStyles.overlay}>
        <div className={popupStyles.popup}>
            <Button variant="outline-danger" onClick = {(e)=> handleDeletePost(props.post.id)}>Delete Post</Button>

            <span className={popupStyles.close} onClick={closeHandler}>
            &times;
            </span>

            <div className={popupStyles.content}>
                {/* {props.children} */}

                <div style = {{height: "10%", width: "95%"}}>
                    <Row>
                        <Col xs={6}>
                            <div style = {{height: "100%", width: "100%"}}>
                                <Form onSubmit={handleUpdate}>
                                    <MediaPreview mediaLink = {props.post.link}/>
                                    <Form.Group className="mb-3" controlId="formBasicDescription">
                                        <Form.Label>
                                            {descInput}
                                        </Form.Label>
                                        <Form.Control name="imageURL" onChange={handleImageURLInput} placeholder="Enter online image url" value={media}/>
                                        <Form.Control as="textarea" onChange={handleDescInput}
                                        placeholder= "Change the text to..."
                                        rows={3}
                                        value={descInput}/>
                                    </Form.Group>
                                    <Button variant="primary" type="submit">
                                        Done
                                    </Button>
                                </Form>
                            </div>
                        </Col>
                        <Col></Col>
                        <Col >
                            <Row> {commentsToast} </Row>
                        </Col>
                    </Row>
                </div>
            </div>
        </div>
        </div>
    );
};

PostEditPopUp.propTypes = {
  title: PropTypes.string,
  show: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired
};

export default PostEditPopUp;