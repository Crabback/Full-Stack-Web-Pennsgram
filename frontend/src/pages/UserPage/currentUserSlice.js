import { createSlice } from '@reduxjs/toolkit'

export const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState: {
    id: -1,
    username: 'NOT_A_USER',
    password: '',
    followings: [],
    followers: [],
    avatar:"https://img.wattpad.com/8f19b412f2223afe4288ed0904120a48b7a38ce1/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f776174747061642d6d656469612d736572766963652f53746f7279496d6167652f5650722d38464e2d744a515349673d3d2d3234323931353831302e313434336539633161633764383437652e6a7067",
    posts: [],
    followerSuggestions: []
  },
  reducers: {
    updateCurrentUser: (state, action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      return action.payload //you must return payload to receive the 
    },
    logoutAction: state => {
      return state = {
        id: -1,
        username: 'NOT_A_USER',
        password: '',
        followings: [],
        followers: [],
        avatar:"https://img.wattpad.com/8f19b412f2223afe4288ed0904120a48b7a38ce1/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f776174747061642d6d656469612d736572766963652f53746f7279496d6167652f5650722d38464e2d744a515349673d3d2d3234323931353831302e313434336539633161633764383437652e6a7067",
        posts: [],
        followerSuggestions: []
      }
    },
  }
})

export const { updateCurrentUser, logoutAction } = currentUserSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`
export const selectCurrentUser = (state) => state.currentUser

export default currentUserSlice.reducer