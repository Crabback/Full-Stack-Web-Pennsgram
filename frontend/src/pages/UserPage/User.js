import "../../Styles.css";
//import { getPost } from "../../api/mock_api";
import React, { useState, useEffect } from "react";
import { Footer } from "../../components/Footer";
import ReactRoundedImage from "react-rounded-image";
import { CardCustomedUserPage } from '../../components/CardUserPage';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Button from 'react-bootstrap/Button';
import { NavLink, useParams, useNavigate } from "react-router-dom";
import { selectCurrentUser, updateCurrentUser } from './currentUserSlice'
import { useSelector, useDispatch } from 'react-redux'
import { getUser, getPost, getPosts, followUser, unfollowUser } from "../../fetcher";
import ToggleButton from 'react-bootstrap/ToggleButton';
import PostEditPopUp from './PostEditPopUp.js'
import InfiniteScroll from "react-infinite-scroll-component";

export default function User() {

  const dispatch = useDispatch();
  const navigate = useNavigate();
  // the logged-in user who is browsing page (self)
  const stateCurrentUser = useSelector(selectCurrentUser);
  const [thisUser, setThisUser] = useState(stateCurrentUser);
  const [allPosts, setThisUserPosts] = useState([]);

  const [thisUserPostsIds, setThisUserPostsIds] = useState(stateCurrentUser.posts.slice(0, 3));
  const [numberPostsShown, setNumberPostsShown] = useState(3);
  const [hasMore, setHasMore] = useState(true);

  const [offset, setOffset] = useState(0);
  const [checked, setChecked] = useState(true);
  const [text, setText] = useState("Follow");
  const [isFollowing, setIsFollowing] = useState(true);

  let { username } = useParams();
  // action trigger states
  const [editedAndRefreshCards, setEditedAndRefreshCards] = useState(true); //the value of this switch doesn't matter only the change matter for triggering useEffect
  const [hearFromDeleteComment, setHearFromDeleteComment] = useState(false); //the request for refreshing comment list from PopUp window, trigger value doesnt matter
  // states for post edit pop up
  const [visibility, setVisibility] = useState(false);
  const [postBeingEdited, setPostBeingEdited] = useState({});

  useEffect(() => {
    async function fetchData() {
      //verification username exist or not
      let output1 = await getUser(username);
      let output2 = await getPosts();
      //get the user object
      if (output1 === null) {
        alert("User Not Found.")
        navigate(`/user/${stateCurrentUser.username}`);
      } else {
        setThisUser(output1);
        setThisUserPostsIds(output1.posts.slice(0, numberPostsShown));
        setThisUserPosts(output2);
      }
    }
    try { fetchData(); }
    catch (err) {
      console.error(err);
    }

    if (stateCurrentUser.followings.includes(username)) {
      setText("Following");
      setChecked(false);
      setOffset(0);
      setIsFollowing(true);
    } else {
      setText("Follow");
      setChecked(true);
      setOffset(0);
      setIsFollowing(false);
    }

    // if hearFromDeleteComment: action -> setPostBeingEdited
    async function updatePostBeingEdited() {
      //refresh the same post with comment deleted
      console.log("before postBeingEdited.id", postBeingEdited.id);
      const updatedPost = await getPost(JSON.stringify(postBeingEdited) !== '{}' ? postBeingEdited.id : -1);
      setPostBeingEdited(updatedPost);
      console.log("after postBeingEdited.id", postBeingEdited.id);

    }
    try {
      if (JSON.stringify(postBeingEdited) !== '{}' && postBeingEdited) { updatePostBeingEdited() };
    } catch (err) {
      console.log(err);
      console.log("Error: User: useEffect: updatePostBeingEdited");
    }

  }, [username, editedAndRefreshCards, hearFromDeleteComment, numberPostsShown]); //adding dependency making sure useEffect only run once after each render


  const popupCloseHandler = () => {
    setVisibility(false);
  };

  const fetchMorePost = () => {
    console.log("thisUser.posts.length: ", thisUser.posts.length);
    if (numberPostsShown >= thisUser.posts.length) {
      setHasMore(false);
      return;
    }
    // a fake async api call like which sends
    // 20 more records in .5 secs
    setTimeout(() => {

      setThisUserPostsIds(thisUser.posts.slice(0, numberPostsShown + 3));
      console.log("numberPostsShown: ", numberPostsShown);
      setNumberPostsShown(n => n + 3);

    }, 1300);
  };

  const posts = (thisUser===null || thisUser.posts.length===0 )?  (<div>No posts yet. Easily uploads </div>) : 
  <div>
  <hr />
  <InfiniteScroll
    dataLength={thisUserPostsIds.length}
    next={fetchMorePost}
    hasMore={hasMore}
    loader={<h4>Loading...</h4>}
    endMessage={
      <p style={{ paddingTop: "2rem", textAlign: "center" }}>
        <b>Yay! You have seen it all</b>
      </p>
    }
  >
    
    {    thisUserPostsIds.map((p)=>(
    <CardCustomedUserPage username={stateCurrentUser.username} post={allPosts.filter(n => n.id === p)[0]} 
                          setEditedAndRefreshCards = {setEditedAndRefreshCards} oldEditedAndRefreshCards={editedAndRefreshCards} 
                          setPostBeingEdited = {setPostBeingEdited} setVisibility= {setVisibility} key={p}/>
  ))}
  </InfiniteScroll>
  </div>

  function ActionButton() {
    // everything except the API call is visual only
    async function handleFollow(e) {
      setChecked(e.currentTarget.checked);
      if (text === "Follow") {
        setText("Following");
        if (isFollowing) { setOffset(0); }
        else { setOffset(1); }
        const updatedUser = await followUser(stateCurrentUser.username, username);
        dispatch(updateCurrentUser(updatedUser));
      } else {
        setText("Follow");
        if (isFollowing) { setOffset(-1); }
        else { setOffset(0); }
        const updatedUser = await unfollowUser(stateCurrentUser.username, username);
        dispatch(updateCurrentUser(updatedUser));
      }
    }

    if (username === stateCurrentUser.username && username !== "NOT_A_USER") {
      return (
        <NavLink to="/upload">
          <Button> New Post </Button>
        </NavLink>
      )
    } else if (stateCurrentUser.username !== "NOT_A_USER") {
      return (
        <ToggleButton className="mb-2" id="toggle-check" type="checkbox" variant="outline-primary"
          data-testid="followingButton" checked={checked} onChange={handleFollow}>
          {text}
        </ToggleButton>
      )
    } else {
      return (
        <NavLink to="/login">
          <Button> Login to Follow </Button>
        </NavLink>
      )
    }
  }


  return (
    <div className='background'>
      <div style={{ display: 'flex', justifyContent: 'center' }}>
        <ReactRoundedImage
          image={thisUser.avatar}
          roundedColor="white"
          imageWidth="150"
          imageHeight="150"
          roundedSize="10"
          borderRadius="100" />
      </div>
      <div style={{ display: 'flex', justifyContent: 'center', paddingTop: "1rem" }}>
        <p className="fw-bold" data-testid="username"> {thisUser.username} </p>
      </div>
      <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
        <Row>
          <Col> <NavLink to={"/followinglist/" + thisUser.username} style={{ textDecoration: "none" }}> {thisUser.followings.length} Following </NavLink> </Col>
          <Col> <NavLink to={"/followerlist/" + thisUser.username} style={{ textDecoration: "none" }}> {thisUser.followers.length + offset} Followers </NavLink> </Col>
        </Row>
      </div>

      <Container>
        <Col sm={{ span: 4, offset: 11 }} style={{ paddingBottom: '2rem', paddingTop: '2rem' }}>
          <ActionButton />
        </Col>
        <Col sm={15} style={{ display: 'flex', justifyContent: 'center' }}>
          <Row> {posts} </Row>
        </Col>
      </Container>

      <PostEditPopUp className="popWindow" post={postBeingEdited}
        setEditedAndRefreshCards={setEditedAndRefreshCards} oldEditedAndRefreshCards={editedAndRefreshCards}
        setHearFromDeleteComment={setHearFromDeleteComment} oldHearFromDeleteComment={hearFromDeleteComment}
        setPostBeingEdited={setPostBeingEdited}
        onClose={popupCloseHandler} show={visibility} >
        <h1>Hello This is Popup Content Area</h1>
        <h2>This is my lorem ipsum text here!</h2>
      </PostEditPopUp>


      <div style={{ paddingLeft: "2rem" }}>
        <Footer />
      </div>
    </div>
  );
}
