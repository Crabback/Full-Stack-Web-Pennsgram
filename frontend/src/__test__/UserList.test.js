import '@testing-library/jest-dom/extend-expect';

import React from "react";
import { getByDisplayValue, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";
import UserList from "../pages/UserList";
import { Provider } from 'react-redux';
import { configureStore } from "@reduxjs/toolkit";
import store from '../Store/store';
import renderer from 'react-test-renderer';
import { getUser } from "../api/mock_api";
import { updateCurrentUser, selectCurrentUser} from '../pages/UserPage/currentUserSlice';



const store1 = store;

test('render following list ', async () => {
    const currentUser = {username: 'obama', password: 'obama123'};
    const userRoster = await getUser(currentUser.username);
    let user;
    userRoster.forEach(element => {
      user = element;
    });
    store1.dispatch(updateCurrentUser(user));
    const { getByText } = render(
        <Provider store={store1}>
          <BrowserRouter>
            <UserList />
          </BrowserRouter>
        </Provider>
    );
    const following = getByText(/zanemao/);
    expect(following).toBeVisible();

});

test('Page matches snapshot', async  () => {

    const component = renderer.create(
        <Provider store={store1}>
        <BrowserRouter>
            <UserList />
        </BrowserRouter>
        </Provider>

    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });


test('render follower list ', async () => {
    const { getByText }  = render(
        <Provider store={store1}>
          <BrowserRouter>
            <UserList list="follower"/>
          </BrowserRouter>
        </Provider>
    );
    const follower = getByText(/trump/);
    expect(follower).toBeVisible();

});



