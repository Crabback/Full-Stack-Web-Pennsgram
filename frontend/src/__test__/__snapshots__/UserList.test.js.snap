// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page matches snapshot 1`] = `
<div
  className="background"
>
  <h2
    style={
      Object {
        "paddingBottom": "1rem",
        "paddingLeft": "2rem",
      }
    }
  >
     
    Following
     
  </h2>
  <ul
    className="list-group"
    onKeyDown={[Function]}
    style={
      Object {
        "paddingBottom": "2rem",
        "paddingLeft": "2rem",
        "paddingRight": "70%",
      }
    }
  >
    <li
      className="list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <a
        className="button_text"
        href="/user/zanemao"
        onClick={[Function]}
      >
        zanemao
      </a>
    </li>
    <li
      className="list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <a
        className="button_text"
        href="/user/elon"
        onClick={[Function]}
      >
        elon
      </a>
    </li>
    <li
      className="list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <a
        className="button_text"
        href="/user/tiger"
        onClick={[Function]}
      >
        tiger
      </a>
    </li>
    <li
      className="list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <a
        className="button_text"
        href="/user/curry"
        onClick={[Function]}
      >
        curry
      </a>
    </li>
    <li
      className="list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <a
        className="button_text"
        href="/user/pig"
        onClick={[Function]}
      >
        pig
      </a>
    </li>
  </ul>
  <a
    className="button_text"
    href="/user/obama"
    onClick={[Function]}
    style={
      Object {
        "paddingLeft": "2rem",
      }
    }
  >
    <button
      className="btn btn-primary"
      disabled={false}
      type="button"
    >
      Back
    </button>
  </a>
  <div
    style={
      Object {
        "paddingLeft": "2rem",
        "paddingTop": "1rem",
      }
    }
  >
    <div
      className="footer"
    >
      <p
        className="footer_elements"
      >
        Terms And Policies
      </p>
      <p
        className="footer_elements"
      >
        All Copy Right Reserved
      </p>
    </div>
  </div>
</div>
`;
